plutus-ledger-api/src/PlutusLedgerApi/Data/V1/Value.hs

```diff
@@ -0,0 +1,134 @@
+
+newtype Value = Value {getValue :: Map.Map CurrencySymbol (Map.Map TokenName Integer)}
+  deriving stock (Haskell.Show)
+  deriving newtype (PlutusTx.ToData, PlutusTx.FromData, PlutusTx.UnsafeFromData)
+
+instance Haskell.Eq Value where
+  (==) = eq
+
+instance Eq Value where
+  {-# INLINEABLE (==) #-}
+  (==) = eq
+
+{-# INLINEABLE unordEqWith #-}
+
+{- | Check equality of two lists given a function checking whether a 'Value' is zero and a function
+checking equality of values.
+
+This function recurses on both the lists in parallel and checks whether the key-value pairs are
+equal pointwise. If there is a mismatch, then it tries to find the left key-value pair in the right
+list. If that succeeds then the pair is removed from both the lists and recursion proceeds pointwise
+as before until there's another mismatch. If at some point a key-value pair from the left list is
+not found in the right one, then the function returns 'False'. If the left list is exhausted, but
+the right one still has some non-zero elements, the function returns 'False' as well.
+
+We check equality of values of two key-value pairs right after ensuring that the keys match. This is
+disadvantageous if the values are big and there's a key that is present in one of the lists but not
+in the other, since in that case computing equality of values was expensive and pointless. However
+
+1. we've checked and on the chain 'Value's very rarely contain 'CurrencySymbol's with more than 3
+   'TokenName's associated with them, so we optimize for the most common use case
+2. computing equality of values before ensuring equality of all the keys certainly does help when we
+   check equality of 'TokenName'-value pairs, since the value of a 'TokenName' is an 'Integer' and
+   @(==) :: Integer -> Integer -> Bool@ is generally much faster than repeatedly searching for keys
+   in a list
+3. having some clever logic for computing equality of values right away in some cases, but not in
+   others would not only complicate the algorithm, but also increase the size of the function and
+   this resource is quite scarce as the size of a program growing beyond what's acceptable by the
+   network can be a real deal breaker, while general performance concerns don't seem to be as
+   pressing
+
+The algorithm we use here is very similar, if not identical, to @valueEqualsValue4@ from
+https://github.com/input-output-hk/plutus/issues/5135
+-}
+unordEqWith ::
+  forall k a.
+  ( PlutusTx.Eq k
+  , PlutusTx.UnsafeFromData k
+  , PlutusTx.UnsafeFromData a
+  , PlutusTx.ToData k
+  , PlutusTx.ToData a
+  ) =>
+  (a -> Bool) ->
+  (a -> a -> Bool) ->
+  Map k a ->
+  Map k a ->
+  Bool
+unordEqWith is0 eqV = goBoth
+  where
+    -- Recurse on the spines of both the lists simultaneously.
+    goBoth kvsL kvsR
+      -- One spine is longer than the other one, but this still can result in a
+      -- succeeding equality.
+      -- check if the non-empty list only contains zero values.
```

Comment:
That was probably my typo, but please remove that dot and the newline after "equality", it's a single sentence.